### ISSUE TITLE
Assertion failed: !more (src/lb.cpp:110)

### ISSUE THREAD

**Issue by kondrenkov** [zeromq/libzmq#1588]
There is a simple DEALER-DEALER communication using multi-part messages.
And ZeroMQ **crashes** if multipart message cannot be delivered atomically.

##### Steps to reproduce:
1) sender(DEALER) binds socket;
2) receiver(DEALER) connects to sender's socket;
3) sender sends 1-st part of multipart message;
4) receiver disconnects from sender;
5) sender tries to send 2-nd part and **crashes**.

##### Test
I've prepared test for this issue (https://github.com/zeromq/issues/pull/15), and the output for several zmq version:
```sh
>>> Ubuntu 14.04.3 ZeroMQ 4(master)
Current ØMQ version is 4.2.0
Assertion failed: !more (src/lb.cpp:110)
Aborted (core dumped)

>>> Ubuntu14.04.3 ZeroMQ 4.1.3
Current ØMQ version is 4.1.3
Assertion failed: !more (src/lb.cpp:110)
Aborted (core dumped)

>>> Ubuntu 12.04.4 ZeroMQ 3.2.4
Current ØMQ version is 3.2.4
Assertion failed: !more (lb.cpp:97)
Aborted (core dumped)
```

**Comment by c-rack** [comment]
@kondrenkov Thanks for submitting this issue and the test case. It was very helpful to reproduce the problem.

I played a little bit with the test and found that we can just remove the `zmq_assert(!more)` without breaking any other test in our test suite. Your test will pass then if you would assert for `-1` on the second `zmq_send` return code.

So maybe we should just remove the `zmq_assert(!more)` line.

**Comment by kondrenkov** [comment]
@c-rack Thank you for you support!
I've just fixed the test https://github.com/kondrenkov/issues/commit/7912b99613732f38cf9983395e393ddca2fe092c .

Now I think that test looks pretty well https://github.com/zeromq/issues/pull/15/files .

**Comment by c-rack** [comment]
Great, do you mind sending a pull request that removes the `zmq_assert(!more)` to solve your issue?
Please either add a `"fixes #1588"` in your pull request / commit message or close this issue after your pull request has been merged.

**Comment by lsniks** [comment]
@c-rack If we remove zmq_assert (), would the atomicity of multipart msg still be guaranteed ? Probably not. Existing applications that are built with notion of atomic delivery of multipart msgs will be broken with this change, correct ? Please correct me If I missed something.

**Comment by c-rack** [comment]
Not sure about the atomicity of multipart messages. If the removal of the `zmq_assert` causes a problem with atomicity that is currently not covered by the test suite, that would be another problem that could be addressed by another issue and pull request. IMHO, removing the assert does not break the build and solves this issue, so that change looks good for me. Please correct me if I am wrong.

**Comment by djelenc** [comment]
I'd say that the atomicity is broken if the receiver receives only part of the message. Can the fix cause this to happen? Specifically, if the receiver is reconnected, will it receive the second part of the message or will the message simply get lost?

**Comment by lsniks** [comment]
@c-rack Per zeromq guide definition http://api.zeromq.org/2-1:zmq-send, 

Multi-part messages
A ØMQ message is composed of 1 or more message parts; each message part is an independent zmq_msg_t in its own right. ØMQ ensures atomic delivery of messages; peers shall receive either all message parts of a message or none at all.

zeromq ensures atomic delivery of messages. While the fix suggested solves issue of assert, it breaks this principle of atomic delivery and hence its really not the fix. The designer chose to assert when atomic delivery was not possible which is one way of making sure all sends are atomic. I think we need to look at this fix differently. If we are removing assert on the send side, then may be we should probably add handling of atomicity of multipart msgs on recv side or not have concept of atomic delivery at all.

Please advise.

**Comment by c-rack** [comment]
Ok, I think what we need is a test case (and implementation to pass it) for the atomicity as described in the API. For @kondrenkov, removing the `zmq_assert` would be a fix for his problem, as it solves his problem (i.e. returning `-1` and `errno = EAGAIN` on `zmq_send` instead of assertion failure).

So if the removal of this line breaks the contract / API without breaking the build, we should rather add a minimal test case that breaks the build and then fix that issue. What do you think?

**Comment by lsniks** [comment]
@c-rack  May be I wasn't clear. zmq_assert (!more) is not a bug. It is called in this place to avoid breaking atomicity when system is under load (HWM is reached). If we remove zmq_assert () we are breaking this feature.

Quoting from above,

"So if the removal of this line breaks the contract / API without breaking the build, we should rather add a minimal test case that breaks the build and then fix that issue. What do you think?"

IMHO, If we know that the fix is going to break a feature may we should fix this differently than trying to go in circles.

In my opinion, there could be following fix for the issue :

1. Remove zmq_assert as suggested above and release note possible scenarios where atomicity is not guaranteed. Can have possible unknown impact to existing applications that expect assert in such scenarios. 
2. Rollback previous parts of the msg. This will guarantee atomicity. Haven't fully explored if this is possible in current design though.

**Comment by c-rack** [comment]
@lsniks Thanks for the clarification!

The problem is that `pipe[current]->write()` returns `false` due to the closed peer and not because of HWM reached. The `check_write` function in `pipe_t` checks both and returns `false` if one of those conditions is true.

As you mentioned, the `zmq_assert` was included to deal with HVM-reached situations, so I just added the following check before the `zmq_assert` to detect a terminated peer if HVM is not reached:

```c++
        if (pipes [current]->check_hwm()) {
          // Write failed, but HVM is ok, so it failed because peer terminated
          pipe_terminated (pipes [current]);
          break;
        }
```
_To be inserted here:_ https://github.com/zeromq/libzmq/blob/master/src/lb.cpp#L109

With this check applied, @kondrenkov's test passes and we do not break the atomicity feature.
What do you think?

**Comment by c-rack** [comment]
Just played a little bit more with this and found my suggestion is not the right solution.

@lsniks do you think this is sufficient to retain the assert behaviour?
```c++
    if (!pipes [current]->check_hwm())
        zmq_assert (!more);
```
This way, the assert would only be checked if the failed write was caused by a reached HVM and not by a closed / terminated peer.

Comments?

**Comment by lsniks** [comment]
@c-rack My apologies for getting back late on this. I have been busy traveling.

This issue is reported in our environment due to HWM and not because of closed/terminated peer. @kondrenkov used this scenario because it was easier to reproduce it this way.

check_hwm will not help. We already call this API in write->check_write->check_hwm before writing into the pipe.

Ideally it is expected behavior from zeromq to assert if atomicity cannot be guaranteed. But I think we can solve it via following instead of asserts :

nikhil@nikhil-VirtualBox:~/lib/src/libzmq$ git diff
diff --git a/src/lb.cpp b/src/lb.cpp
index 44b11ea..08b0a04 100644
--- a/src/lb.cpp
+++ b/src/lb.cpp
@@ -107,7 +107,14 @@ int zmq::lb_t::sendpipe (msg_t *msg_, pipe_t **pipe_)
             break;
         }
 
-        zmq_assert (!more);
+        if (more)
+        {
+            pipes [current]->rollback ();
+            more = 0;
+            errno = EAGAIN;
+            return -1;
+        }
+
         active--;
         if (current < active)
             pipes.swap (current, active);
nikhil@nikhil-VirtualBox:~/lib/src/libzmq$ 

Not entirely sure of the impact of rollback, it being a relatively newer API. Please advise.

**Comment by c-rack** [comment]
Thanks, looks good to me. Do you want to submit a pull request?

**Comment by lsniks** [comment]
Sure. Just added.

https://github.com/zeromq/libzmq/pull/1600

Please review/accept.

**Comment by c-rack** [comment]
Thanks!

Closing this issue after PR https://github.com/zeromq/libzmq/pull/1600 was merged.

**Comment by kondrenkov** [comment]
@c-rack 
What about testcase for this issue? https://github.com/zeromq/issues/pull/15

**Comment by c-rack** [comment]
Sorry, maybe I have closed this issue too fast. Does the PR solve the issue? I reopen this.

**Comment by lsniks** [comment]
@c-rack Any updates on this ? Do you see any issues ? I see @hintjens has merged this pull request into master as well.

Please let me know.

**Comment by c-rack** [comment]
@lsniks No, I do not see any issue. Your changes are perfectly ok and also @kondrenkov's testcase works without any problems on current master on my machine. That is why I have closed this issue after the merge.

@kondrenkov: Do you see any issues? Can we close this issue?

**Comment by c-rack** [comment]
@kondrenkov: if your question was about the status of your PR https://github.com/zeromq/issues/pull/15, someone else has to merge it, as I do not have the proper rights to merge it.

**Comment by xaqq** [comment]
Okay I have merged  zeromq/issues#15 since from what i've read it works fine.

**Comment by kondrenkov** [comment]
@c-rack  Thank you.
No I haven't any issues for this point of time.
I'd like to close this ticket.
